### ISSUE TITLE
Grid.ChildrenTransitions are not always animated when using States

### ISSUE THREAD

**Issue by sonnemaf** [microsoft/microsoft-ui-xaml#324]
**Describe the bug**
I have created a Style/Template which makes a CheckBox look like an ToggleButton. Don't ask me why I just did. The Template contains a blue (AccentColor) slider Rectangle which is moved to a different column using a Setter in the Checked states. The Grid has a RepositionThemeTransition in the ChildrenTransitions. This should animate the Column change. This works in DEBUG mode (F5) but not if you don't use the debugger (Ctrl+F5).

![image](https://user-images.githubusercontent.com/9335559/52942486-62959c00-336b-11e9-89dc-0b669e12523a.png)

On my page I have two CheckBoxes (Sound and Video), both with this Style. The IsChecked property of the Video CheckBox is databound (x:Bind) to the Sound CheckBox.

**Steps to reproduce the bug**
1. Take this code

```xml 
<Page x:Class="App35.MainPage"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:local="using:App35"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      mc:Ignorable="d"
      RequestedTheme="Dark"
      Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
    <Page.Resources>
        <Style x:Key="IPhoneCheckBoxStyle"
               TargetType="CheckBox">
            <Setter Property="Background"
                    Value="{ThemeResource CheckBoxBackgroundUnchecked}" />
            <Setter Property="Foreground"
                    Value="{ThemeResource CheckBoxForegroundUnchecked}" />
            <Setter Property="BorderBrush"
                    Value="{ThemeResource CheckBoxBorderBrushUnchecked}" />
            <Setter Property="Padding"
                    Value="8,5,0,0" />
            <Setter Property="HorizontalAlignment"
                    Value="Left" />
            <Setter Property="VerticalAlignment"
                    Value="Center" />
            <Setter Property="HorizontalContentAlignment"
                    Value="Left" />
            <Setter Property="VerticalContentAlignment"
                    Value="Top" />
            <Setter Property="FontFamily"
                    Value="{ThemeResource ContentControlThemeFontFamily}" />
            <Setter Property="FontSize"
                    Value="{ThemeResource ControlContentThemeFontSize}" />
            <Setter Property="MinWidth"
                    Value="120" />
            <Setter Property="MinHeight"
                    Value="32" />
            <Setter Property="UseSystemFocusVisuals"
                    Value="{StaticResource UseSystemFocusVisuals}" />
            <Setter Property="FocusVisualMargin"
                    Value="-3" />
            <Setter Property="Template">
                <Setter.Value>
                    <ControlTemplate TargetType="CheckBox">
                        <Grid x:Name="RootGrid">
                            <Grid.ChildrenTransitions>
                                <TransitionCollection>
                                    <RepositionThemeTransition />
                                </TransitionCollection>
                            </Grid.ChildrenTransitions>
                            <Grid.ColumnDefinitions>
                                <ColumnDefinition Width="50" />
                                <ColumnDefinition Width="50" />
                                <ColumnDefinition />
                            </Grid.ColumnDefinitions>
                            <VisualStateManager.VisualStateGroups>
                                <VisualStateGroup x:Name="CombinedStates">
                                    <VisualState x:Name="UncheckedNormal">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundUnchecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundUnchecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushUnchecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="UncheckedPointerOver">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundUncheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundUncheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushUncheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="UncheckedPressed">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundUncheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundUncheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushUncheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="UncheckedDisabled">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundUncheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundUncheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushUncheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="CheckedNormal">
                                        <VisualState.Setters>
                                            <Setter Target="SliderRectangle.(Grid.Column)"
                                                    Value="1" />
                                        </VisualState.Setters>
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundChecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundChecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushChecked}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="CheckedPointerOver">
                                        <VisualState.Setters>
                                            <Setter Target="SliderRectangle.(Grid.Column)"
                                                    Value="1" />
                                        </VisualState.Setters>
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundCheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundCheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushCheckedPointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="CheckedPressed">
                                        <VisualState.Setters>
                                            <Setter Target="SliderRectangle.(Grid.Column)"
                                                    Value="1" />
                                        </VisualState.Setters>
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundCheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundCheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushCheckedPressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="CheckedDisabled">
                                        <VisualState.Setters>
                                            <Setter Target="SliderRectangle.(Grid.Column)"
                                                    Value="1" />
                                        </VisualState.Setters>
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundCheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundCheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushCheckedDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="IndeterminateNormal">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundIndeterminate}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundIndeterminate}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushIndeterminate}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="IndeterminatePointerOver">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundIndeterminatePointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundIndeterminatePointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushIndeterminatePointerOver}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="IndeterminatePressed">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundIndeterminatePressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundIndeterminatePressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushIndeterminatePressed}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                    <VisualState x:Name="IndeterminateDisabled">
                                        <Storyboard>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ContentPresenter"
                                                                           Storyboard.TargetProperty="Foreground">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxForegroundIndeterminateDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="Background">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBackgroundIndeterminateDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="RootGrid"
                                                                           Storyboard.TargetProperty="BorderBrush">
                                                <DiscreteObjectKeyFrame KeyTime="0"
                                                                        Value="{ThemeResource CheckBoxBorderBrushIndeterminateDisabled}" />
                                            </ObjectAnimationUsingKeyFrames>
                                        </Storyboard>
                                    </VisualState>
                                </VisualStateGroup>
                            </VisualStateManager.VisualStateGroups>
                            <ContentPresenter x:Name="ContentPresenter"
                                              Grid.Column="2"
                                              Margin="8,0"
                                              HorizontalAlignment="Left"
                                              VerticalAlignment="Center" />
                            <Rectangle Fill="{ThemeResource SystemControlForegroundBaseHighBrush}"
                                       Grid.ColumnSpan="2" />
                            <TextBlock Text="On"
                                       Foreground="{ThemeResource SystemColorButtonTextBrush}"
                                       Margin="8,0"
                                       HorizontalAlignment="Left"
                                       VerticalAlignment="Center" />
                            <TextBlock Grid.Column="1"
                                       Text="Off"
                                       Foreground="{ThemeResource SystemColorButtonTextBrush}"
                                       Margin="8,0"
                                       HorizontalAlignment="Right"
                                       VerticalAlignment="Center" />
                            <Rectangle x:Name="SliderRectangle"
                                       Grid.Column="0"
                                       Margin="3"
                                       Fill="{ThemeResource SystemControlBackgroundAccentBrush}" />
                        </Grid>
                    </ControlTemplate>
                </Setter.Value>
            </Setter>
        </Style>
    </Page.Resources>

    <Grid>
        <CheckBox Content="Sound"
                  x:Name="SoundCheckBox"
                  Style="{StaticResource IPhoneCheckBoxStyle}"
                  HorizontalAlignment="Left"
                  Height="37"
                  Margin="159,110,0,0"
                  VerticalAlignment="Top"
                  Width="335" />
        <CheckBox Content="Video"
                  IsChecked="{x:Bind SoundCheckBox.IsChecked, Mode=OneWay}"
                  Style="{StaticResource IPhoneCheckBoxStyle}"
                  HorizontalAlignment="Left"
                  Height="35"
                  Margin="159,177,0,0"
                  VerticalAlignment="Top"
                  Width="323" />

    </Grid>
</Page>
```

2. Run the app and tap/click the Sound CheckBox. 

**Expected behavior**
The first time the slider of the Sound CheckBox should slide with an animation. This only works the second or third time. The Video CheckBox is not animated untill you tap/click it. 

When you start the app and only use the keyboard (TAP and SPACE keys) it will never animate.

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)                   |    |
| October 2018 Update (17763)   | Yes   |
| April 2018 Update (17134)        | Yes   |
| Fall Creators Update (16299)     | Yes   |
| Creators Update (15063)           |    |
| Anniversary Update (14393)      |    |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes   |
| Mobile                   |    |
| Xbox                      |    |
| Surface Hub          |     |
| IoT                         |     |

**Additional context**

**Comment by jevansaks** [comment]
@stmoy is there a more reliable solution that facades offers here?

**Comment by stmoy** [comment]
Hi @sonnemaf ! It sounds to me like there is a bug in RepositionThemeTransition if it's only animating sometimes but not other times. I will follow up with the team.

Regarding facades/simple animatable properties - you can potentially leverage TranslationTransition to automatically animate changes to the Translation property on UIElement. A few caveats: the Translation property and TranslationTransition are only available in the October 2018 update (and newer), and the Translation property needs to be set in codebehind. 

You can view an example of TranslationTransition in the [Xaml Controls Gallery app](https://aka.ms/xamlgalleryapp) under Motion > Implicit Transitions.

**Comment by sonnemaf** [comment]
The composition animation alternative is not an option for me. I really need it in a Style/Template. I have been using this technique since Windows 8.0 and just recently noticed this bug while creating this demo for a UWP App development Training. Might have been there for a while though.

Thank you for checking this with the team.

**Comment by gegao18** [comment]
Hi @sonnemaf, I debugged your repro and it's a general bug with Xaml's ThemeTransitions.

In this case, since the target of the transition is a Rectangle element created from a control template, Xaml doesn't fully create the element. There are parts that are created lazily as they're needed. ThemeTransitions happen to check whether one of these parts exists, but doesn't try to create it if it doesn't exist. The RepositionThemeTransition doesn't find anything and early exits without starting the animation.

The workaround is to use a feature that creates the part that ThemeTransition is looking for. Putting an event handler on the element is one way to do this. I got the RepositionThemeTransition playing consistently in your repro by adding an empty Loaded event handler:

                            <Rectangle x:Name="SliderRectangle"
                                       Loaded="SliderRectangle_Loaded"
                                       Grid.Column="0"
                                       Margin="3"
                                       Fill="{ThemeResource SystemControlBackgroundAccentBrush}" />

The body of the event handler itself doesn't matter and can be empty. Just the fact that it's listening for an event is enough.
